---
title: "Seurat - Guided Clustering Tutorial"
output:
  html_document:
    theme: united
    df_print: kable
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
***

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 120),
  message = FALSE,
  warning = FALSE
)
```


```{r option-and-library}
rm(list=ls())
options(stringsAsFactors = F)
library("Seurat")
library("dplyr") # At least group_by is called from this package.
library("Matrix") # provides formal class 'dgCMatrix', representing sparse Matrix
library("readr") # provides fast read_tsv to read the large table.
library("magrittr") # provides snazzy pipeline operation
```



```{r load-data}
initial_clustering <- TRUE

favorite_project_name <- "GSE_scRNA_seq"
input_file <- file.path("../data", favorite_project_name, "all_data.txt")

metadata_attributes <- c("sample_names",
                         "processed_by_Maxima_enzyme",
                         "Lymph_node",
                         "classified_as_cancer_cell",
                         "classified_as_non-cancer_cells",
                         "non-cancer_cell_type")

metadata_list <- list()

for (i in seq_along(metadata_attributes)){
  print(i)
  metadata_list[[metadata_attributes[i]]] <- 
    scan(file = input_file, skip = i -1, nlines = 1, sep = "\t", what = "")[-1]
}

raw_expression_matrix <- 
  read_tsv(file = input_file, col_names = F, skip = length(metadata_attributes)) %>%
  data.frame(., row.names = 1) %>%
  as.data.frame(.) %>%
  set_colnames(., metadata_list[["sample_names"]]) %>%
  set_rownames(., gsub("'", "", rownames(.))) %>%
  data.matrix(.) %>%
  Matrix(., sparse = TRUE)


if (initial_clustering == TRUE) {
  input_expression_matrix <- raw_expression_matrix
} else {
  # Re-cluster need to subset the cells, namely remove some columns.

clustered_cells <- readRDS(file = "../output/cluster_identities.rds")

cells_to_recluster <- names(clustered_cells)[clustered_cells %in% c(0, 1)]
retained_column_index <- colnames(raw_expression_matrix) %in% cells_to_recluster

some_cells_expression_matrix <- raw_expression_matrix[, retained_column_index]
  
input_expression_matrix <- some_cells_expression_matrix
}

mydata <- CreateSeuratObject(counts = input_expression_matrix,
                           min.cells = 3,
                           min.features = 200,
                           project = favorite_project_name)
# object.size(expression_matrix)

mydata

first_ID <- sapply(colnames(input_expression_matrix), function(x) strsplit(x, "_")[[1]][1])
table(first_ID)
```
```{r}
lymph_nodes <- metadata_list[["Lymph_node"]]
sample_names <- metadata_list[["sample_names"]]
combined_lymph_samples <- paste(lymph_nodes, sample_names, sep = "-")
first_lymph_ID <- sapply(combined_lymph_samples, function(x) strsplit(x, "_")[[1]][1])
table(first_lymph_ID)
length(table(first_lymph_ID))
```



```{r mito, fig.height=7, fig.width=13}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
mydata[["percent.mt"]] <- PercentageFeatureSet(object = mydata, pattern = "^MT-")
sum(mydata[["percent.mt"]])
```

```{r qc1, fig.height=7, fig.width=13}
# Show QC metrics for the first 5 cells
head(x = mydata@meta.data, 10)
```



```{r qc2, fig.height=7, fig.width=13}

#Visualize QC metrics as a violin plot
VlnPlot(object = mydata, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(object = mydata, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plot2 <- FeatureScatter(object = mydata, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
CombinePlots(plots = list(plot1,plot2))

mydata <- subset(x = mydata, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```


```{r normalize}
# mydata <- NormalizeData(object = mydata, normalization.method = "LogNormalize", scale.factor = 1e4)
# mydata[["RNA"]]@data
```


```{r var-features, fig.height=5, fig.width=11}
mydata <- FindVariableFeatures(object = mydata, selection.method = 'vst', nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(x = VariableFeatures(object = mydata), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(object = mydata)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))
```

```{r regressvarmt, fig.height=7, fig.width=11, results='hide',eval = FALSE}
mydata <- ScaleData(object = mydata, vars.to.regress = 'percent.mt')
```


### Perform linear dimensional reduction

Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using `features` argument if you wish to choose a different subset.

```{r pca,results='hide'}
mydata <- RunPCA(object = mydata, features = VariableFeatures(object = mydata))
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including `VizDimReduction`, `DimPlot`, and `DimHeatmap`

```{r pca_viz, message=TRUE}
# Examine and visualize PCA results a few different ways
print(x = mydata[['pca']], dims = 1:5, nfeatures = 5)
VizDimLoadings(object = mydata, dims = 1:2, reduction = 'pca')
DimPlot(object = mydata, reduction = 'pca')
```

In particular `DimHeatmap` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Setting `cells` to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.

```{r single-heatmap}
DimHeatmap(object = mydata, dims = 1, cells = 500, balanced = TRUE)
```

```{r multi-heatmap, fig.height=12, fig.width=9}
DimHeatmap(object = mydata, dims = 1:15, cells = 500, balanced = TRUE)
```


### Determine the 'dimensionality' of the dataset

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many componenets should we choose to include? 10? 20? 100?

In [Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8), we implemented a resampling test inspired by the JackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a 'null distribution' of feature scores, and repeat this procedure. We identify 'significant' PCs as those who have a strong enrichment of low p-value features.

```{r jackstraw, fig.height=6, fig.width=10}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More approximate techniques such as those implemented in ElbowPlot() can be used to reduce computation time
mydata <- JackStraw(object = mydata, num.replicate = 100)
mydata <- ScoreJackStraw(object = mydata, dims = 1:20)
```

The `JackStrawPlot` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 'Significant' PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.

```{r jsplots, fig.height=6, fig.width=10}
JackStrawPlot(object = mydata, dims = 1:15)
```

An alternative heuristic method generates an 'Elbow plot': a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot` function). In this example, we can observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs. 

```{r elbow_plot, fig.height=6, fig.width=10}
ElbowPlot(object = mydata)
```

Identifying the true dimensionality of a dataset -- can be challenging/uncertain for the user. We therefore suggest these three approaches to consider. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. The third is a heuristic that is commonly used, and can be calculated instantly. In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-12 as a cutoff. 

We chose 10 here, but encourage users to consider the following:

* Dendritic cell and NK aficionados may recognize that genes strongly associated with PCs 12 and 13 define rare immune subsets (i.e. MZB1 is a marker for plasmacytoid DCs). However, these groups are so rare, they are difficult to distinguish from background noise for a dataset of this size without prior knowledge. 
* We encourage users to repeat downstream analyses with a different number of PCs (10, 15, or even 50!). As you will observe, the results often do not differ dramatically.
* We advise users to err on the higher side when choosing this parameter. For example, performing downstream analyses with only 5 PCs does signifcanltly and adversely affect results.

***

```{r cluster, fig.height=5, fig.width=7}
mydata <- FindNeighbors(object = mydata, dims = 1:10)
mydata <- FindClusters(object = mydata, resolution = 0.5)

# Look at cluster IDs of the first 5 cells
head(x = Idents(object = mydata), 5)
# saveRDS(Idents(object = mydata), file = "../output/cluster_identities.rds")
```

***

### Run non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{r tsne, fig.height=5, fig.width=7}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages = "umap-learn")
mydata <- RunUMAP(object = mydata, dims = 1:10)
```

```{r tsneplot, fig.height=5, fig.width=7}
# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
DimPlot(object = mydata, label = TRUE, reduction = 'umap')
```

You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.

```{r saveobject, eval=FALSE}
saveRDS(mydata, file = "../output/pbmc_tutorial.rds")
```

***

### Finding differentially expressed features (cluster biomarkers)

Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells.  `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The `min.pct` argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, `max.cells.per.ident` can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.

```{r markers1, fig.height=8, fig.width=15}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(object = mydata, ident.1 = 1, min.pct = 0.25)
cluster1.markers
head(x = cluster1.markers, n = 5)
# find all markers distinguishing cluster 5 from clusters 0 and 3
# cluster5.markers <- FindMarkers(object = mydata, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
# head(x = cluster5.markers, n = 5)
# find markers for every cluster compared to all remaining cells, report only the positive ones
mydata.markers <- FindAllMarkers(object = mydata, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
# all markers can be saved to a local file to see the genes.
mydata.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

Seurat has several tests for differential expression which can be set with the test.use parameter (see our [DE vignette](http://satijalab01.nygenome.org/seurat/v3.0/de_vignette.html) for details). For example, the ROC test returns the 'classification power' for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r markersroc, fig.height=8, fig.width=15}
cluster1.markers <- FindMarkers(object = mydata, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster1.markers
```

We include several tools for visualizing marker expression. `VlnPlot` (shows expression probability distributions across clusters), and `FeaturePlot` (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring `RidgePlot`, `CellScatter`, and `DotPlot` as additional methods to view your dataset.

```{r markerplots, fig.height=8, fig.width=15}
VlnPlot(object = mydata, features = c("MS4A1", "CD79A"))
# you can plot raw counts as well
VlnPlot(object = mydata, features = c("NKG7", "PF4"), slot = 'counts', log = TRUE)
FeaturePlot(object = mydata, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))
```

`DoHeatmap` generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.

```{r clusterHeatmap, fig.height=8, fig.width=15}
mydata.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC) -> top10
DoHeatmap(object = mydata, features = top10$gene) + NoLegend()
```

### Assigning cell type identity to clusters

Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:

Cluster ID | Markers       | Cell Type
-----------|---------------|----------
0          | IL7R, CCR7    | Memory CD4+ 
1          | CD14, LYZ     | CD14+ Mono
2          | IL7R, S100A4  | Naive CD4+ T 
3          | MS4A1         | B 
4          | CD8A          | CD8+ T 
5          | FCGR3A, MS4A7 | FCGR3A+ Mono
6          | GNLY, NKG7    | NK 
7          | FCER1A, CST3  | DC
8          | PPBP          | Mk


```{r labelplot, fig.height=5, fig.width=9}
new.cluster.ids <- c("Memory CD4 T", "CD14+ Mono", "Naive CD4 T", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Mk")
names(x = new.cluster.ids) <- levels(x = mydata)
mydata <- RenameIdents(object = mydata, new.cluster.ids)
DimPlot(object = mydata, reduction = 'umap', label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r save.img, include=FALSE}
library(ggplot2)
plot <- DimPlot(object = mydata, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") + 
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + 
  guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "../output/images/pbmc3k_umap.png", height = 7, width = 12, plot = plot)
```

```{r save.rds}
saveRDS(mydata, file = "../output/pbmc3k_final.rds")
```
